### ISSUE TITLE
Testing error with HttpTestingController, inconsistent types

### ISSUE THREAD

**Issue by davidwalschots** [angular/angular#18345]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[x] Bug report  
[ ] Feature request
[x] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
</code></pre>

## Current behavior
The [documentation](https://angular.io/guide/http#error-handling) of the `HttpClient` states:
> Also, if something goes wrong client-side, such as an exception gets thrown in an RxJS operator, or if a network error prevents the request from completing successfully, an actual `Error` will be thrown.

For which the following example is provided:
```
(err: HttpErrorResponse) => {
  if (err.error instanceof Error) {
    // A client-side or network error occurred. Handle it accordingly.
    console.log('An error occurred:', err.error.message);
  }
}
```
The `error` function of `TestRequest` requires passing an `ErrorEvent`, which is then the resulting error type found within `HttpErrorResponse.error`. Thus, I'm unable to simulate client-side or network errors in testing.

```
httpMock.expectOne(`/api/test`).error(new ErrorEvent('error')); //Error type instance can't be passed.
```

## Expected behavior
Either the `TestRequest.error` function accepts an `Error` type as input, or the documentation on `HttpErrorResponse` handling is updated to include the `ErrorEvent` type as one of the options (which still leaves actual `Error` type testing to be desired).
## Environment

Angular version: 4.3.1

**Comment by alxhub** [comment]
`ErrorEvent` is the correct type - it will always be an `ErrorEvent` and not an `Error` as far as I know.

**Comment by scottasmith** [comment]
I have been having problems with this following the documentation.
Looking at the code, the ErrorEvent is the correct error type and it comes directly from XHR. 

https://github.com/angular/angular/blob/6e8e3bd2488eb0896932fe1a10afefbb0dccbca9/packages/common/http/src/xhr.ts#L232

So looks like the docs need updating

**Comment by acwatson** [comment]
Here is an example of something that seems to be **impossible** to unit test. My scenario is that I want to catch a timeout error when an http request times out.

**Code that makes the request:**

return this._http.get('/api/login', {headers: headers})
      .timeout(10000)
      .map((responseData) => {
        return LoginAttemptStatus.SUCCESS;
      })
      .catch(err => {
        if ('TimeoutError' === err.name) {
          return Observable.of(LoginAttemptStatus.TIMEOUT);
        }
      });

**Here is an attempt to set up a unit test with the HttpTestingController to throw a TimeoutError:**

testingController.expectOne((req: HttpRequest<any>) => {
      return req.method === 'GET';
    }).error(new ErrorEvent('TimeoutError', new TimeoutError()));

**Problem**

When I run my unit test, the object supplied to my catch block is of type HttpErrorResponse, but in real life, Rxjs (5.5.2) will send a TimeoutError object to the catch block. So, it appears that there is no way I can use HttpTestingController to mock getting a TimeoutError to be received in my catch block.

Note - using the older MockBackend class, I was able to get this to work like so:
mockBackend.connections.subscribe(connection => {
      connection.mockError(new TimeoutError());
    });

**Comment by scottasmith** [comment]
Not impossible. There is a way but involves more code. 

The [```HttpClientTestingModule```](https://github.com/angular/angular/blob/6e8e3bd2488eb0896932fe1a10afefbb0dccbca9/packages/common/http/testing/src/modules) that is imported into the TestBed provides [```HttpClientTestingBackend```](https://github.com/angular/angular/blob/6e8e3bd2488eb0896932fe1a10afefbb0dccbca9/packages/common/http/testing/src/backend.ts)

You could write your own version of ```HttpClientTestingBackend``` and ```TestRequest``` and provide that in the TestBed

**Comment by nponiros** [comment]
This might not be the best issue to mention this since my issue is not really related to testing, but here it goes. I can open another issue if wanted, since there are some more docs issues when it comes to HttpClient error handling.

The docs were updated to use `ErrorEvent` instead of `Error` but apparently `onerror` actually has a `ProgressEvent` as parameter and not an `ErrorEvent`. There is a TypeScript issue about this (https://github.com/Microsoft/TypeScript/issues/21414). The XMLHttpRequest spec also suggests it should be a `ProgressEvent` (https://xhr.spec.whatwg.org/#event-xhr-error).

Testing on Chrome, Safari and Firefox I also got an instance of `ProgressEvent` when I had no connection to the server and not an `ErrorEvent`. Other Browsers might use an `ErrorEvent`.
